apps/aecore/src/aec_fork_signalling.erl

```diff
@@ -0,0 +1,350 @@
+%%%
+%%% The block at height HE - 1 is the last signalling block, and there must
+%%% be a result computed for it. Once a worker process that computes the
+%%% result is spawned, a 'pending' atom for the given block hash is  stored in
+%%% results map. Once the worker process is done with the computation it will
+%%% send a boolean result and the 'pending' atom is replaced with the boolean
+%%% result.
+%%%
+%%% The blocks with height between HE and FH (inclusive) are treated
+%%% differently. Once such block arrives, its previous key hash is looked up
+%%% in the results map. If it's present, it means that the previous key hash
+%%% belongs to the block at height HE - 1 (which has result set to 'true',
+%%% 'false' or 'pending'), or the previous key hash has a pointer (key block
+%%% hash) to a block at height HE - 1.
+%%%
+%%% The pointers are used because the result at height HE - 1 can change
+%%% asynchronosly (from 'pending' to a boolean value). Moreover, it's not
+%%% necessary to search a block at HE - 1 height to get the result in case
+%%% a new block arrives. Instead, the results map may have an entry for
+%%% previous key hash pointing to the block with HE - 1 height.
+%%%
+%%% The worst case scenario is restarting the node at height H2 - 1. If the
+%%% node has the top key block at height H2 - 1 and the results map is empty,
+%%% it needs to start a worker process the will search for a block at HE - 1,
+%%% and then it will start computing the fork result. Some components of the
+%%% system may be waiting for the final result.
+%%%
+%%% @end
+%%%============================================================================
+-module(aec_fork_signalling).
+
+-behaviour(gen_server).
+
+%% API
+-export([start_link/0,
+         get_fork_result/2,
+         compute_fork_result/2
+        ]).
+
+%% gen_server callbacks
+-export([init/1,
+         handle_call/3,
+         handle_cast/2,
+         handle_info/2,
+         terminate/2
+        ]).
+
+-export([worker_process/4]).
+
+-define(SERVER, ?MODULE).
+
+-type header_hash()      :: aec_blocks:block_header_hash().
+
+-type worker()           :: {pid(), reference()}.
+
+-type result()           :: true | false | pending.
+
+%% The pointer is a key block hash which belongs to the last signalling block.
+-type pointer()          :: header_hash().
+
+-record(state, {fork     :: aec_hard_forks:fork() | undefined,
+                results  :: #{header_hash() := result() | pointer()},
+                workers  :: #{worker() := header_hash()}
+               }).
+
+-define(IS_RESULT(X), X =:= true orelse X =:= false orelse X =:= pending).
+-define(IS_POINTER(X), is_binary(X)).
+
+%% API
+
+start_link() ->
+    gen_server:start_link({local, ?SERVER}, ?MODULE, [], []).
+
+get_fork_result(Block, BlockHash) ->
+    gen_server:call(?SERVER, {get_fork_result, Block, BlockHash}).
+
+compute_fork_result(Block, BlockHash) ->
+    gen_server:cast(?SERVER, {compute_fork_result, Block, BlockHash}).
```

Comment:
Does this mean that on node from an empty storage, the aec_conductor process floods this aec_fork_signalling  process? 🤔

Reply:
You're probably right. There should be a case in the conductor which checks the origin of the block, if it's database, then skip. I'll add a TODO in the code.

Reply:
Ah, I misread the comment, you meant an empty storage. I will have a look at it later.

Reply:
User for the first time starts node. All PoWs are all computed so blocks get retrieved from network and added to storage quickly. This cast is asynchronous so I am concerned for message queue length. It feels some state of the process may need to go in (publicly readable, privately writable?) ETS table so to perform check quickly in context of conductor.

Reply:
The agreed solution is to move the fork config from the `aec_fork_signalling` to `aec_conductor` and make block height vs signalling end/fork height check in the context of the conductor and pass fork config as a param from the conductor.

Reply:
Fixed in 12b7f6d07.